### ISSUE TITLE
FIPS support incompatible with OpenSSL 3.0

### ISSUE THREAD

**Issue by sharkcz** [opencryptoki/libica#58]
There are some changes in the FIPS support coming with OpenSSL 3.0 (https://wiki.openssl.org/index.php/OpenSSL_3.0#Upgrading_from_the_OpenSSL_2.0_FIPS_Object_Module) that break libica builds.
````
...
fips.c: In function 'fips_init':
fips.c:121:17: warning: implicit declaration of function 'FIPS_mode_set'; did you mean 'FIPS_mode'? [-Wimplicit-function-declaration]
  121 |                 FIPS_mode_set(1);
      |                 ^~~~~~~~~~~~~
      |                 FIPS_mode
...
libtool: link: gcc -Wall -Wextra -mzarch -DICA_FIPS -O3 -D_FORTIFY_SOURCE=2 -I./include -I./../include -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -march=z13 -mtune=z14 -fasynchronous-unwind-tables -fstack-clash-protection -Wl,-z -Wl,relro -Wl,--as-needed -Wl,-z -Wl,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -o .libs/icainfo icainfo-icainfo.o  ./.libs/libica.so -lrt -lcrypto -ldl -pthread
make[1]: Leaving directory '/builddir/build/BUILD/libica-3.7.0/src'
/usr/bin/ld: ./.libs/libica.so: undefined reference to `U64'
/usr/bin/ld: ./.libs/libica.so: undefined reference to `FIPS_mode_set'
collect2: error: ld returned 1 exit status
````
Please see https://bugzilla.redhat.com/show_bug.cgi?id=1952946 for all details.

**Comment by jschmidb** [comment]
Hi Dan, thanks for the hint! I'm not sure if we can provide a fix for our upcoming May shipment, but this is definitely a point to look at.

**Comment by sharkcz** [comment]
I think I could temporarily make a build without the FIPS support, so we won't have an issue with libica in RHEL-9 Beta. And then apply a proper fix when available.

**Comment by jschmidb** [comment]
Agreed. OpenSSL 3.0 is not yet on our radar, but we are aware that we will have to make some adaptions.

**Comment by sharkcz** [comment]
@jschmidb , could you take a look, please?

**Comment by jschmidb** [comment]
Hi Dan, yes, it's on my list!

**Comment by jschmidb** [comment]
Just to make sure we are talking about the same thing:  in issue #62 you wrote:

> The gnutls rpm package does the same thing as openssl above (modifies the "post-install" hook for rpmbuild), so libica will have to use the same mechanism. Looking at the gnutls build.log they create hmac during the build using the upstream provided rule and then recreate it using the rpmbuild hook. So you shouldn't be changing anything and I will update the rpm package. The only update that would be useful is to provide a wrapper over the openssl command to create the hmac, both openssl and gnutls seems to have one. Or we could require the hmaccalc package to be installed, which is what the kernel does.

This is what I still kept in mind ...

**Comment by jschmidb** [comment]
Fixed in libica 4.0.0
Closing.
